File: faq/app/text_utils.py
from __future__ import annotations

import hashlib
import html
import re
from typing import Iterable


_MD_IMAGE_RE = re.compile(r"!\[([^\]]*)\]\(([^)]+)\)")
_HTML_TAG_RE = re.compile(r"<[^>]+>")
_PUNCT_RE = re.compile(r"[\s\W_]+", re.UNICODE)


def stable_hash(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def strip_markdown(text: str) -> str:
    """Convert answer markdown/html to compact plain text for indexing."""
    s = text or ""
    s = _MD_IMAGE_RE.sub(lambda m: f"{m.group(1)} {m.group(2)}", s)
    s = re.sub(r"`([^`]+)`", r"\1", s)
    s = re.sub(r"[*_#>|~\-]+", " ", s)
    s = _HTML_TAG_RE.sub(" ", s)
    s = html.unescape(s)
    s = re.sub(r"\s+", " ", s)
    return s.strip()


def answer_summary(answer: str, max_chars: int = 900) -> str:
    text = strip_markdown(answer)
    if len(text) <= max_chars:
        return text
    return text[: max_chars - 1].rstrip() + "…"


def normalize_query(text: str) -> str:
    return re.sub(r"\s+", " ", (text or "").strip())


def char_ngrams(text: str, min_n: int = 2, max_n: int = 3) -> list[str]:
    cleaned = _PUNCT_RE.sub("", (text or "").lower())
    tokens: list[str] = []
    for n in range(min_n, max_n + 1):
        if len(cleaned) < n:
            continue
        tokens.extend(cleaned[i : i + n] for i in range(0, len(cleaned) - n + 1))
    # Add alnum words so English model/menu names still work.
    tokens.extend(re.findall(r"[a-z0-9]+(?:-[a-z0-9]+)*", (text or "").lower()))
    return tokens


def keyword_text(texts: Iterable[str]) -> str:
    toks: list[str] = []
    for text in texts:
        toks.extend(char_ngrams(text))
    return " ".join(toks)


def clip_text(text: str, limit: int) -> str:
    s = (text or "").strip()
    if len(s) <= limit:
        return s
    return s[: limit - 1].rstrip() + "…"

File: Router/server/src/services/rag/textUtils.js
import crypto from "node:crypto";

const MD_IMAGE_RE = /!\[([^\]]*)\]\(([^)]+)\)/g;
const HTML_TAG_RE = /<[^>]+>/g;
const PUNCT_RE = /[\s\W_]+/gu;

function htmlUnescape(s) {
    return s
        .replace(/&amp;/g, "&")
        .replace(/&lt;/g, "<")
        .replace(/&gt;/g, ">")
        .replace(/&quot;/g, '"')
        .replace(/&#039;/g, "'")
        .replace(/&#(\d+);/g, (_, n) => String.fromCharCode(Number(n)));
}

export function stableHash(text) {
    return crypto.createHash("sha256").update(text || "", "utf-8").digest("hex");
}

export function stripMarkdown(text) {
    let s = text || "";
    s = s.replace(MD_IMAGE_RE, (_, alt, src) => `${alt} ${src}`);
    s = s.replace(/`([^`]+)`/g, "$1");
    s = s.replace(/[*_#>|~\-]+/g, " ");
    s = s.replace(HTML_TAG_RE, " ");
    s = htmlUnescape(s);
    s = s.replace(/\s+/g, " ");
    return s.trim();
}

export function answerSummary(answer, maxChars = 900) {
    const text = stripMarkdown(answer);
    if (text.length <= maxChars)
        return text;
    return `${text.slice(0, maxChars - 1).trimEnd()}…`;
}

export function normalizeQuery(text) {
    return (text || "").trim().replace(/\s+/g, " ");
}

export function charNgrams(text, minN = 2, maxN = 3) {
    const cleaned = (text || "").toLowerCase().replace(PUNCT_RE, "");
    const tokens = [];
    for (let n = minN; n <= maxN; n++) {
        if (cleaned.length < n)
            continue;
        for (let i = 0; i <= cleaned.length - n; i++) {
            tokens.push(cleaned.slice(i, i + n));
        }
    }
    const words = (text || "").toLowerCase().match(/[a-z0-9]+(?:-[a-z0-9]+)*/g) ?? [];
    tokens.push(...words);
    return tokens;
}

export function keywordText(texts) {
    const toks = [];
    for (const text of texts) {
        toks.push(...charNgrams(text));
    }
    return toks.join(" ");
}

export function clipText(text, limit) {
    const s = (text || "").trim();
    if (s.length <= limit)
        return s;
    return `${s.slice(0, limit - 1).trimEnd()}…`;
}

export function msSince(t0) {
    return Math.round((performance.now() - t0) * 10) / 10;
}
